### ISSUE TITLE
[request] LLVM Benchmarks

### ISSUE THREAD

**Issue by fosskers** [lehins/massiv#21]
Repa warns that performance can go out the window if we're not careful. One thing it advertises to help with this is to compile with `-fllvm`. In my old repa-based benchmarks for `mapalgebra`, I generally saw a 2x speedup. Do you know how LLVM affects Massiv?

Regardless of the answer here, I'll be testing both for `mapalgebra`. I plan to write up a long blog post about it with all sorts of benchmarks, and turning on LLVM would be included with that.

**Comment by lehins** [comment]
Honestly speaking I haven't done anything even close to extensive benchmarking with LLVM, but from what I have tried there was a slight speedup. Which is great, it means that users of massiv don't need to  rely on some heavy compiler to get all the performance they deserve. :)

From what I noticed, proportionally Repa has a bigger gain from LLVM than massiv, but not enough to get to performance of massiv anyways. It's easy to understand why authors of Repa were advertising`-fllvm` almost as a requirement, they heavily relied on optimizations available only with llvm, which they go into in their papers about repa, eg.:
> This in turn enables LLVM’s global value numbering transformation, which recovers sharing of intermediate results between the computation of successive array elements.

One thing I would highly recommend is avoiding use of lists inside stencils, otherwise, I suspect, it might destroy all of the optimizations like loop unrolling and cache friendly traversal, but I might be wrong. In any case, let me know how will the  benchmark go and I might be able to suggest a thing or two on how to speed it up. It will be as much of an experiment as it is for you ;)

Looking forward to reading your blogpost and anxious to see the results of benchmarks.
